### PR TITLE
Adding Main Menu with Measure Heart Rate & Export HR data

### PR DESCRIPTION
**Pull request by sanketshah92**
- XML design for Main Menu
- Added Navigation Component for navigation
- Added Custom drawables for backgroud setup of menu item
- Added utility for scaling top visible item and scale down other menu items
- Moved MainActivity to presentation package

### CODE REVIEW

File: app/src/main/java/com/sanket/watchapplication/presentation/mainMenu/utils/CustomScrollingLayoutCallback.kt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+            val centerOffset = height.toFloat() / 2.0f / parent.height.toFloat()
+            val yRelativeToCenterOffset = y / parent.height + centerOffset
+
+            // Normalize for center
+            progressToCenter = abs(0.5f - yRelativeToCenterOffset)
+            // Adjust to the maximum scale
+            progressToCenter = progressToCenter.coerceAtMost(MAX_ICON_PROGRESS)
+
+            scaleX = 1 - progressToCenter
+            scaleY = 1 - progressToCenter
```
Comment (EntyR) [inline app/src/main/java/com/sanket/watchapplication/presentation/mainMenu/utils/CustomScrollingLayoutCallback.kt]: No need for scaling in this project, just light up center element

**Review by EntyR — CHANGES_REQUESTED** [review]

**Comment by EntyR** [comment]
image

**Top and bottom elements is cut off**

**Comment by EntyR** [comment]
Please add snap helper to center central element

**Comment by EntyR** [comment]
Also please compare ui design with one from task document. It must look similar

**Comment by sanketshah92** [comment]
> Please add snap helper to center central element

Okay, Adding it in upcoming commit.

**Comment by sanketshah92** [comment]
@EntyR I have updated the UI and changed scale behaviour to selector.

**Comment by sanketshah92** [comment]
device-2022-05-25-180115

**Comment by EntyR** [comment]
@sanketshah92 

Design template doesn't categories it's properly so I'll put it here 

1st template its **normal** state and the rest 2 it's **pressed** state.
 
Also please watch for spacing between elements and for their size to fit design. 

Spacing is a little bit big and elements are small

image

**Comment by EntyR** [comment]
image


1. Please lighten up the **central element**. Can change transparency of non-central for example

2. Also look for elements **size on larger devices**. Only 3 element should be on the screen at the time. Look if you can use [different dimen resources](https://developer.android.com/guide/topics/large-screens/support-different-screen-sizes) or calculate size with screen height and width

File: app/src/main/java/com/sanket/watchapplication/presentation/mainMenu/MainMenuFragment.kt [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+                    }
+                }
+            })
+        setupMenuView()
+    }
+
+    private fun setupMenuView() {
+        val snapHelper = LinearSnapHelper()
+        snapHelper.attachToRecyclerView(mainMenuRecycler)
+        mainMenuRecycler.apply {
```
Comment (EntyR) [inline app/src/main/java/com/sanket/watchapplication/presentation/mainMenu/MainMenuFragment.kt:56]: Please [set up view binding](https://developer.android.com/topic/libraries/view-binding) and use it to access ui elements instead

File: app/src/main/res/drawable/rounded_bg_main_menu.xml [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+        </shape>
+    </item>
+    <item
+        android:bottom="4dp"
+        android:left="4dp"
+        android:right="4dp"
+        android:top="4dp">
+        <shape android:shape="rectangle">
+            <solid android:color="#5F006F" />
+            <corners android:radius="32dp" />
```
Comment (EntyR) [inline app/src/main/res/drawable/rounded_bg_main_menu.xml]: Same as above please extract all resource into color and dimen files

File: app/src/main/res/drawable/rounded_bg_focused_menu_item.xml [hunk 4]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<shape xmlns:android="http://schemas.android.com/apk/res/android"
+    android:shape="rectangle">
+    <stroke
+        android:width="2dp"
+        android:color="#5F006F" />
+    <corners android:radius="32dp" />
```
Comment (EntyR) [inline app/src/main/res/drawable/rounded_bg_focused_menu_item.xml]: Please extract all resource into color and dimen files

File: app/src/main/res/layout/list_item_main_menu.xml [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:layout_marginStart="8dp"
+    android:layout_marginTop="6dp"
+    android:layout_marginEnd="8dp"
+    android:background="@drawable/selector_main_menu_item_press"
+    android:fontFamily="sans-serif-condensed-medium"
+    android:gravity="center"
+    android:minHeight="56dp"
+    android:text="Action 1"
```
Comment (EntyR) [inline app/src/main/res/layout/list_item_main_menu.xml]: Same as above please extract all resource into color and dimen files

File: app/src/main/java/com/sanket/watchapplication/presentation/mainMenu/MainMenuFragment.kt [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+        prepareAdapter()
+    }
+
+    private fun getMenuData(): List<String> {
+        val menuItems = mutableListOf<String>()
+        menuItems.add("Action 1")
+        menuItems.add("Measure HR")
+        menuItems.add("Export HR Data")
+        menuItems.add("Action 3")
+        menuItems.add("Action 4")
```
Comment (EntyR) [inline app/src/main/java/com/sanket/watchapplication/presentation/mainMenu/MainMenuFragment.kt]: There is no need for Action n elements anymore. Also please extract all string values

**Review by EntyR — CHANGES_REQUESTED** [review]

File: app/build.gradle [hunk 7]
```diff
@@ -38,4 +38,7 @@ dependencies {
     implementation 'androidx.navigation:navigation-fragment-ktx:2.4.2'
     implementation 'androidx.navigation:navigation-ui-ktx:2.4.2'
     implementation 'com.airbnb.android:lottie:5.1.1'
+    implementation 'com.intuit.sdp:sdp-android:1.0.6'
+    implementation 'com.intuit.ssp:ssp-android:1.0.6'
```
Comment (EntyR) [inline app/build.gradle:42]: Lets try to avoid usage of third-party libraries

**Review by EntyR — CHANGES_REQUESTED** [review]

**Comment by EntyR** [comment]
Also please answer comments if the problem has been fixed. Something like "done" or "fixed" is enough

File: app/src/main/java/com/sanket/watchapplication/presentation/mainMenu/MainMenuFragment.kt (hunk 6 above)
Comment (sanketshah92) [reply app/src/main/java/com/sanket/watchapplication/presentation/mainMenu/MainMenuFragment.kt]: done

File: app/src/main/res/layout/list_item_main_menu.xml (hunk 5 above)
Comment (sanketshah92) [reply app/src/main/res/layout/list_item_main_menu.xml]: done

File: app/src/main/res/drawable/rounded_bg_focused_menu_item.xml (hunk 4 above)
Comment (sanketshah92) [reply app/src/main/res/drawable/rounded_bg_focused_menu_item.xml]: done

File: app/src/main/res/drawable/rounded_bg_main_menu.xml (hunk 3 above)
Comment (sanketshah92) [reply app/src/main/res/drawable/rounded_bg_main_menu.xml]: extracted

**Comment by sanketshah92** [comment]
> @sanketshah92
> 
> Design template doesn't categories it's properly so I'll put it here
> 
> 1st template its **normal** state and the rest 2 it's **pressed** state.
> 
> Also please watch for spacing between elements and for their size to fit design.
> 
> Spacing is a little bit big and elements are small
> 
> image



> Also please answer comments if the problem has been fixed. Something like "done" or "fixed" is enough

can we use these libs for this assignment? As managing more screen dimensions will be bit time consuming, core functionalities also need to be addressed.

**Comment by EntyR** [comment]
> can we use these libs for this assignment? As managing more screen dimensions will be bit time consuming, core functionalities also need to be addressed.

We should avoid using any 3-rd party libraries. In some place it's necessary like in heart rate screen, but if you can not use it you better don't

**Review by EntyR — APPROVED** [review]

**Pull request merged**